File: backend/Services/ShiftService.cs
using backend.Data;
using backend.DTOs;
using Microsoft.EntityFrameworkCore;

namespace backend.Services;

/// <summary>
/// Service xử lý các nghiệp vụ liên quan đến Ca trực và Phân công ca
/// </summary>
public class ShiftService : IShiftService
{
    private readonly SportContext _context;
    private readonly ILogger<ShiftService> _logger;

    public ShiftService(SportContext context, ILogger<ShiftService> logger)
    {
        _context = context;
        _logger = logger;
    }

    #region Shift Management

    public async Task<List<ShiftDto>> GetAllShiftsAsync()
    {
        try
        {
            var shifts = await _context.CaTrucs
                .Include(ct => ct.MaCoSoNavigation)
                .ToListAsync();

            return shifts
                .OrderBy(ct => ct.Ngay)
                .ThenBy(ct => ct.GioBatDau)
                .Select(ct => new ShiftDto
                {
                    MaCa = ct.MaCa,
                    MaCoSo = ct.MaCoSo ?? 0,
                    TenCoSo = ct.MaCoSoNavigation != null ? ct.MaCoSoNavigation.TenCoSo : null,
                    Ngay = ct.Ngay.HasValue ? ct.Ngay.Value.ToDateTime(TimeOnly.MinValue) : default,
                    GioBatDau = ct.GioBatDau.HasValue ? ct.GioBatDau.Value.ToTimeSpan() : default,
                    GioKetThuc = ct.GioKetThuc.HasValue ? ct.GioKetThuc.Value.ToTimeSpan() : default,
                    TenCa = ct.TenCa
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting all shifts: {ex.Message}");
            throw;
        }
    }

    public async Task<ShiftDto?> GetShiftByIdAsync(int maCa)
    {
        try
        {
            var shift = await _context.CaTrucs
                .Include(ct => ct.MaCoSoNavigation)
                .Where(ct => ct.MaCa == maCa)
                .Select(ct => new ShiftDto
                {
                    MaCa = ct.MaCa,
                    MaCoSo = ct.MaCoSo ?? 0,
                    TenCoSo = ct.MaCoSoNavigation != null ? ct.MaCoSoNavigation.TenCoSo : null,
                    Ngay = ct.Ngay.HasValue ? ct.Ngay.Value.ToDateTime(TimeOnly.MinValue) : default,
                    GioBatDau = ct.GioBatDau.HasValue ? ct.GioBatDau.Value.ToTimeSpan() : default,
                    GioKetThuc = ct.GioKetThuc.HasValue ? ct.GioKetThuc.Value.ToTimeSpan() : default,
                    TenCa = ct.TenCa
                })
                .FirstOrDefaultAsync();

            return shift;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting shift {maCa}: {ex.Message}");
            throw;
        }
    }

    public async Task<List<ShiftDto>> GetShiftsByCoSoAsync(int maCoSo)
    {
        try
        {
            var shifts = await _context.CaTrucs
                .Include(ct => ct.MaCoSoNavigation)
                .Where(ct => ct.MaCoSo == maCoSo)
                .ToListAsync();

            return shifts
                .OrderBy(ct => ct.Ngay)
                .ThenBy(ct => ct.GioBatDau)
                .Select(ct => new ShiftDto
                {
                    MaCa = ct.MaCa,
                    MaCoSo = ct.MaCoSo ?? 0,
                    TenCoSo = ct.MaCoSoNavigation != null ? ct.MaCoSoNavigation.TenCoSo : null,
                    Ngay = ct.Ngay.HasValue ? ct.Ngay.Value.ToDateTime(TimeOnly.MinValue) : default,
                    GioBatDau = ct.GioBatDau.HasValue ? ct.GioBatDau.Value.ToTimeSpan() : default,
                    GioKetThuc = ct.GioKetThuc.HasValue ? ct.GioKetThuc.Value.ToTimeSpan() : default,
                    TenCa = ct.TenCa
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting shifts for facility {maCoSo}: {ex.Message}");
            throw;
        }
    }

    public async Task<List<ShiftDto>> GetShiftsByDateAsync(DateTime date)
    {
        try
        {
            var shifts = await _context.CaTrucs
                .Include(ct => ct.MaCoSoNavigation)
                .Where(ct => ct.Ngay.HasValue && ct.Ngay.Value == DateOnly.FromDateTime(date))
                .ToListAsync();

            return shifts
                .OrderBy(ct => ct.GioBatDau)
                .Select(ct => new ShiftDto
                {
                    MaCa = ct.MaCa,
                    MaCoSo = ct.MaCoSo ?? 0,
                    TenCoSo = ct.MaCoSoNavigation != null ? ct.MaCoSoNavigation.TenCoSo : null,
                    Ngay = ct.Ngay.HasValue ? ct.Ngay.Value.ToDateTime(TimeOnly.MinValue) : default,
                    GioBatDau = ct.GioBatDau.HasValue ? ct.GioBatDau.Value.ToTimeSpan() : default,
                    GioKetThuc = ct.GioKetThuc.HasValue ? ct.GioKetThuc.Value.ToTimeSpan() : default,
                    TenCa = ct.TenCa
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting shifts for date {date}: {ex.Message}");
            throw;
        }
    }

    public async Task<int> CreateShiftAsync(CreateShiftRequest request)
    {
        try
        {
            // Kiểm tra cơ sở có tồn tại không
            var coSo = await _context.CoSos.FindAsync(request.MaCoSo);
            if (coSo == null)
            {
                throw new InvalidOperationException("Cơ sở không tồn tại");
            }

            // Kiểm tra giờ hợp lệ
            if (request.GioBatDau >= request.GioKetThuc)
            {
                throw new InvalidOperationException("Giờ bắt đầu phải nhỏ hơn giờ kết thúc");
            }

            // Tạo mã ca mới
            var maxMaCa = await _context.CaTrucs.MaxAsync(ct => (int?)ct.MaCa) ?? 0;
            var newMaCa = maxMaCa + 1;

            var shift = new CaTruc
            {
                MaCa = newMaCa,
                MaCoSo = request.MaCoSo,
                Ngay = DateOnly.FromDateTime(request.Ngay),
                GioBatDau = TimeOnly.FromTimeSpan(request.GioBatDau),
                GioKetThuc = TimeOnly.FromTimeSpan(request.GioKetThuc),
                TenCa = request.TenCa
            };

            _context.CaTrucs.Add(shift);
            await _context.SaveChangesAsync();

            _logger.LogInformation($"Shift created successfully with ID: {newMaCa}");
            return newMaCa;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error creating shift: {ex.Message}");
            throw;
        }
    }

    public async Task<bool> UpdateShiftAsync(int maCa, UpdateShiftRequest request)
    {
        try
        {
            var shift = await _context.CaTrucs.FindAsync(maCa);
            if (shift == null)
            {
                throw new InvalidOperationException("Ca trực không tồn tại");
            }

            // Cập nhật các trường nếu có giá trị mới
            if (request.Ngay.HasValue)
                shift.Ngay = DateOnly.FromDateTime(request.Ngay.Value);


            if (request.GioBatDau.HasValue)
                shift.GioBatDau = TimeOnly.FromTimeSpan(request.GioBatDau.Value);

            if (request.GioKetThuc.HasValue)
                shift.GioKetThuc = TimeOnly.FromTimeSpan(request.GioKetThuc.Value);


            // Kiểm tra giờ hợp lệ sau khi cập nhật
            if (shift.GioBatDau >= shift.GioKetThuc)
            {
                throw new InvalidOperationException("Giờ bắt đầu phải nhỏ hơn giờ kết thúc");
            }

            if (!string.IsNullOrEmpty(request.TenCa))
                shift.TenCa = request.TenCa;

            await _context.SaveChangesAsync();

            _logger.LogInformation($"Shift {maCa} updated successfully");
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error updating shift {maCa}: {ex.Message}");
            throw;
        }
    }

    public async Task<bool> DeleteShiftAsync(int maCa)
    {
        using var transaction = await _context.Database.BeginTransactionAsync();
        try
        {
            var shift = await _context.CaTrucs.FindAsync(maCa);
            if (shift == null)
            {
                throw new InvalidOperationException("Ca trực không tồn tại");
            }

            // Xóa các phân công ca liên quan
            var assignments = await _context.PhanCongCas.Where(pc => pc.MaCa == maCa).ToListAsync();
            _context.PhanCongCas.RemoveRange(assignments);

            // Xóa ca trực
            _context.CaTrucs.Remove(shift);
            await _context.SaveChangesAsync();

            await transaction.CommitAsync();

            _logger.LogInformation($"Shift {maCa} deleted successfully");
            return true;
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError($"Error deleting shift {maCa}: {ex.Message}");
            throw;
        }
    }

    #endregion

    #region Shift Assignment

    public async Task<List<ShiftAssignmentDto>> GetAllShiftAssignmentsAsync()
    {
        try
        {
            var assignments = await _context.PhanCongCas
                .Include(pc => pc.MaNvNavigation)
                .Include(pc => pc.MaCaNavigation)
                .ThenInclude(ca => ca.MaCoSoNavigation)
                .ToListAsync();

            return assignments
                .OrderBy(pc => pc.MaCaNavigation?.Ngay)
                .ThenBy(pc => pc.MaCaNavigation?.GioBatDau)
                .Select(pc => new ShiftAssignmentDto
                {
                    MaPc = pc.MaPc,
                    MaNv = pc.MaNv ?? 0,
                    TenNhanVien = pc.MaNvNavigation != null ? pc.MaNvNavigation.HoTen : null,
                    ChucVu = pc.MaNvNavigation != null ? pc.MaNvNavigation.ChucVu : null,
                    MaCa = pc.MaCa ?? 0,
                    TenCa = pc.MaCaNavigation != null ? pc.MaCaNavigation.TenCa : null,
                    Ngay = pc.MaCaNavigation != null && pc.MaCaNavigation.Ngay.HasValue
                        ? pc.MaCaNavigation.Ngay.Value.ToDateTime(TimeOnly.MinValue)
                        : null,
                    GioBatDau = pc.MaCaNavigation != null && pc.MaCaNavigation.GioBatDau.HasValue
                        ? pc.MaCaNavigation.GioBatDau.Value.ToTimeSpan()
                        : null,
                    GioKetThuc = pc.MaCaNavigation != null && pc.MaCaNavigation.GioKetThuc.HasValue
                        ? pc.MaCaNavigation.GioKetThuc.Value.ToTimeSpan()
                        : null,
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting all shift assignments: {ex.Message}");
            throw;
        }
    }

    public async Task<List<ShiftAssignmentDto>> GetShiftAssignmentsByEmployeeAsync(int maNv)
    {
        try
        {
            var assignments = await _context.PhanCongCas
                .Include(pc => pc.MaNvNavigation)
                .Include(pc => pc.MaCaNavigation)
                .Where(pc => pc.MaNv == maNv)
                .ToListAsync();

            return assignments
                .OrderBy(pc => pc.MaCaNavigation?.Ngay)
                .ThenBy(pc => pc.MaCaNavigation?.GioBatDau)
                .Select(pc => new ShiftAssignmentDto
                {
                    MaPc = pc.MaPc,
                    MaNv = pc.MaNv ?? 0,
                    TenNhanVien = pc.MaNvNavigation != null ? pc.MaNvNavigation.HoTen : null,
                    ChucVu = pc.MaNvNavigation != null ? pc.MaNvNavigation.ChucVu : null,
                    MaCa = pc.MaCa ?? 0,
                    TenCa = pc.MaCaNavigation != null ? pc.MaCaNavigation.TenCa : null,
                    Ngay = pc.MaCaNavigation != null && pc.MaCaNavigation.Ngay.HasValue
                        ? pc.MaCaNavigation.Ngay.Value.ToDateTime(TimeOnly.MinValue)
                        : null,
                    GioBatDau = pc.MaCaNavigation != null && pc.MaCaNavigation.GioBatDau.HasValue
                        ? pc.MaCaNavigation.GioBatDau.Value.ToTimeSpan()
                        : null,
                    GioKetThuc = pc.MaCaNavigation != null && pc.MaCaNavigation.GioKetThuc.HasValue
                        ? pc.MaCaNavigation.GioKetThuc.Value.ToTimeSpan()
                        : null,
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting shift assignments for employee {maNv}: {ex.Message}");
            throw;
        }
    }

    public async Task<List<ShiftAssignmentDto>> GetShiftAssignmentsByShiftAsync(int maCa)
    {
        try
        {
            var assignments = await _context.PhanCongCas
                .Include(pc => pc.MaNvNavigation)
                .Include(pc => pc.MaCaNavigation)
                .Where(pc => pc.MaCa == maCa)
                .ToListAsync();

            return assignments
                .Select(pc => new ShiftAssignmentDto
                {
                    MaPc = pc.MaPc,
                    MaNv = pc.MaNv ?? 0,
                    TenNhanVien = pc.MaNvNavigation != null ? pc.MaNvNavigation.HoTen : null,
                    ChucVu = pc.MaNvNavigation != null ? pc.MaNvNavigation.ChucVu : null,
                    MaCa = pc.MaCa ?? 0,
                    TenCa = pc.MaCaNavigation != null ? pc.MaCaNavigation.TenCa : null,
                    Ngay = pc.MaCaNavigation != null && pc.MaCaNavigation.Ngay.HasValue
                        ? pc.MaCaNavigation.Ngay.Value.ToDateTime(TimeOnly.MinValue)
                        : null,
                    GioBatDau = pc.MaCaNavigation != null && pc.MaCaNavigation.GioBatDau.HasValue
                        ? pc.MaCaNavigation.GioBatDau.Value.ToTimeSpan()
                        : null,
                    GioKetThuc = pc.MaCaNavigation != null && pc.MaCaNavigation.GioKetThuc.HasValue
                        ? pc.MaCaNavigation.GioKetThuc.Value.ToTimeSpan()
                        : null,
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting shift assignments for shift {maCa}: {ex.Message}");
            throw;
        }
    }

    public async Task<int> AssignShiftAsync(AssignShiftRequest request)
    {
        try
        {
            // Kiểm tra nhân viên có tồn tại không
            var employee = await _context.NhanViens.FindAsync(request.MaNv);
            if (employee == null)
            {
                throw new InvalidOperationException("Nhân viên không tồn tại");
            }

            // Kiểm tra ca trực có tồn tại không
            var shift = await _context.CaTrucs.FindAsync(request.MaCa);
            if (shift == null)
            {
                throw new InvalidOperationException("Ca trực không tồn tại");
            }

            // Kiểm tra nhân viên đã được phân công ca này chưa
            var existingAssignment = await _context.PhanCongCas
                .FirstOrDefaultAsync(pc => pc.MaNv == request.MaNv && pc.MaCa == request.MaCa);

            if (existingAssignment != null)
            {
                throw new InvalidOperationException("Nhân viên đã được phân công ca này");
            }

            // Tạo mã phân công mới
            var maxMaPc = await _context.PhanCongCas.MaxAsync(pc => (int?)pc.MaPc) ?? 0;
            var newMaPc = maxMaPc + 1;

            var assignment = new PhanCongCa
            {
                MaPc = newMaPc,
                MaNv = request.MaNv,
                MaCa = request.MaCa
            };

            _context.PhanCongCas.Add(assignment);
            await _context.SaveChangesAsync();

            _logger.LogInformation($"Shift assignment created successfully with ID: {newMaPc}");
            return newMaPc;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error assigning shift: {ex.Message}");
            throw;
        }
    }

    public async Task<bool> RemoveShiftAssignmentAsync(int maPc)
    {
        try
        {
            var assignment = await _context.PhanCongCas.FindAsync(maPc);
            if (assignment == null)
            {
                throw new InvalidOperationException("Phân công ca không tồn tại");
            }

            _context.PhanCongCas.Remove(assignment);
            await _context.SaveChangesAsync();

            _logger.LogInformation($"Shift assignment {maPc} removed successfully");
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error removing shift assignment {maPc}: {ex.Message}");
            throw;
        }
    }

    #endregion
}

File: frontend/src/services/ShiftService.ts
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL;

// Interfaces cho DTOs
export interface ShiftDto {
  maCa: number;
  maCoSo: number;
  tenCoSo?: string;
  ngay: string;
  gioBatDau: string;
  gioKetThuc: string;
  tenCa?: string;
}

export interface CreateShiftRequest {
  maCoSo: number;
  ngay: string;
  gioBatDau: string;
  gioKetThuc: string;
  tenCa: string;
}

export interface UpdateShiftRequest {
  ngay?: string;
  gioBatDau?: string;
  gioKetThuc?: string;
  tenCa?: string;
}

export interface ShiftAssignmentDto {
  maPc: number;
  maNv: number;
  tenNhanVien?: string;
  chucVu?: string;
  maCa: number;
  tenCa?: string;
  ngay?: string;
  gioBatDau?: string;
  gioKetThuc?: string;
}

export interface AssignShiftRequest {
  maNv: number;
  maCa: number;
}

class ShiftService {
  /**
   * Lấy token từ localStorage
   */
  private getToken(): string {
    return localStorage.getItem('access_token') || '';
  }

  /**
   * Tạo config với Authorization header
   */
  private getConfig() {
    return {
      headers: {
        Authorization: `Bearer ${this.getToken()}`,
      },
    };
  }

  // ========== Shift Management ==========

  /**
   * Lấy danh sách tất cả ca trực
   */
  async getAllShifts(): Promise<ShiftDto[]> {
    try {
      const response = await axios.get<ShiftDto[]>(
        `${API_BASE_URL}/shifts`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách ca trực thất bại');
      }
      throw error;
    }
  }

  /**
   * Lấy thông tin ca trực theo mã
   * @param maCa Mã ca trực
   */
  async getShiftById(maCa: number): Promise<ShiftDto> {
    try {
      const response = await axios.get<ShiftDto>(
        `${API_BASE_URL}/shifts/${maCa}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy thông tin ca trực thất bại');
      }
      throw error;
    }
  }

  /**
   * Lấy danh sách ca trực theo cơ sở
   * @param maCoSo Mã cơ sở
   */
  async getShiftsByFacility(maCoSo: number): Promise<ShiftDto[]> {
    try {
      const response = await axios.get<ShiftDto[]>(
        `${API_BASE_URL}/shifts/facility/${maCoSo}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách ca trực thất bại');
      }
      throw error;
    }
  }

  /**
   * Lấy danh sách ca trực theo ngày
   * @param date Ngày (YYYY-MM-DD)
   */
  async getShiftsByDate(date: string): Promise<ShiftDto[]> {
    try {
      const response = await axios.get<ShiftDto[]>(
        `${API_BASE_URL}/shifts/date/${date}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách ca trực thất bại');
      }
      throw error;
    }
  }

  /**
   * Tạo mới ca trực
   * @param request Thông tin ca trực mới
   */
  async createShift(request: CreateShiftRequest): Promise<{ message: string; maCa: number }> {
    try {
      const response = await axios.post<{ message: string; maCa: number }>(
        `${API_BASE_URL}/shifts`,
        request,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Tạo ca trực thất bại');
      }
      throw error;
    }
  }

  /**
   * Cập nhật thông tin ca trực
   * @param maCa Mã ca trực
   * @param request Thông tin cập nhật
   */
  async updateShift(maCa: number, request: UpdateShiftRequest): Promise<{ message: string }> {
    try {
      const response = await axios.put<{ message: string }>(
        `${API_BASE_URL}/shifts/${maCa}`,
        request,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Cập nhật ca trực thất bại');
      }
      throw error;
    }
  }

  /**
   * Xóa ca trực
   * @param maCa Mã ca trực
   */
  async deleteShift(maCa: number): Promise<{ message: string }> {
    try {
      const response = await axios.delete<{ message: string }>(
        `${API_BASE_URL}/shifts/${maCa}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Xóa ca trực thất bại');
      }
      throw error;
    }
  }

  // ========== Shift Assignment ==========

  /**
   * Lấy danh sách tất cả phân công ca
   */
  async getAllShiftAssignments(): Promise<ShiftAssignmentDto[]> {
    try {
      const response = await axios.get<ShiftAssignmentDto[]>(
        `${API_BASE_URL}/shifts/assignments`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách phân công ca thất bại');
      }
      throw error;
    }
  }

  /**
   * Lấy danh sách phân công ca theo nhân viên
   * @param maNv Mã nhân viên
   */
  async getShiftAssignmentsByEmployee(maNv: number): Promise<ShiftAssignmentDto[]> {
    try {
      const response = await axios.get<ShiftAssignmentDto[]>(
        `${API_BASE_URL}/shifts/assignments/employee/${maNv}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách phân công ca thất bại');
      }
      throw error;
    }
  }

  /**
   * Lấy danh sách phân công ca theo ca trực
   * @param maCa Mã ca trực
   */
  async getShiftAssignmentsByShift(maCa: number): Promise<ShiftAssignmentDto[]> {
    try {
      const response = await axios.get<ShiftAssignmentDto[]>(
        `${API_BASE_URL}/shifts/assignments/shift/${maCa}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách phân công ca thất bại');
      }
      throw error;
    }
  }

  /**
   * Phân công ca trực cho nhân viên
   * @param request Thông tin phân công ca
   */
  async assignShift(request: AssignShiftRequest): Promise<{ message: string; maPc: number }> {
    try {
      const response = await axios.post<{ message: string; maPc: number }>(
        `${API_BASE_URL}/shifts/assignments`,
        request,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Phân công ca thất bại');
      }
      throw error;
    }
  }

  /**
   * Xóa phân công ca
   * @param maPc Mã phân công
   */
  async removeShiftAssignment(maPc: number): Promise<{ message: string }> {
    try {
      const response = await axios.delete<{ message: string }>(
        `${API_BASE_URL}/shifts/assignments/${maPc}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Xóa phân công ca thất bại');
      }
      throw error;
    }
  }
}

export default new ShiftService();
